tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,580 @@
+
+  for (auto& ptr : ptrs) {
+    HIP_CHECK(hipHostMalloc(&ptr, allocSize));
+  }
+
+  std::vector<std::thread> threads;
+
+  for (auto ptr : ptrs) {
+    threads.push_back(
+        std::thread([ptr]() { HIP_CHECK_THREAD(workIsDoneCheck<TestType*>(ptr, HostFree)); }));
+  }
+
+  for (auto& t : threads) {
+    t.join();
+  }
+  HIP_CHECK_THREAD_FINALIZE();
+}
+
+#if HT_NVIDIA
+TEMPLATE_TEST_CASE("Unit_hipFreeMultiTArray", "", char, int, float2, float4) {
+  constexpr size_t numAllocs = 10;
+  size_t width = GENERATE(32, 128, 256, 512, 1024);
+  size_t height = GENERATE(32, 128, 256, 512, 1024);
+  DriverContext ctx;
+
+
+  SECTION("ArrayDestroy") {
+    std::vector<hiparray> ptrs(numAllocs);
+    HIP_ARRAY_DESCRIPTOR cuDesc;
+    cuDesc.Width = width;
+    cuDesc.Height = height;
+    cuDesc.Format = HIP_AD_FORMAT_UNSIGNED_INT8;
+    cuDesc.NumChannels = 2;
+    for (auto& ptr : ptrs) {
+      HIP_CHECK(hipArrayCreate(&ptr, &cuDesc));
+    }
+
+    std::vector<std::thread> threads;
+
+    for (auto& ptr : ptrs) {
+      SECTION("ArrayDestroy") {
+        threads.push_back(std::thread(
+            [ptr]() { HIP_CHECK_THREAD(workIsDoneCheck<hiparray>(ptr, ArrayDestroy)); }));
+      }
+    }
+    for (auto& t : threads) {
+      t.join();
+    }
+    HIP_CHECK_THREAD_FINALIZE();
+  };
+
+  SECTION("ArrayFree") {
+    constexpr size_t numAllocs = 10;
+    std::vector<hipArray_t> ptrs(numAllocs);
+    hipExtent extent{};
+    extent.width = width;
+    extent.height = height;
+    hipChannelFormatDesc desc = hipCreateChannelDesc<TestType>();
+
+    for (auto& ptr : ptrs) {
+      HIP_CHECK(hipMallocArray(&ptr, &desc, extent.width, extent.height, hipArrayDefault));
+    }
+
+    std::vector<std::thread> threads;
+
+    for (auto ptr : ptrs) {
+      SECTION("ArrayFree") {
+        threads.push_back(std::thread(
+            [ptr]() { HIP_CHECK_THREAD(workIsDoneCheck<hipArray_t>(ptr, ArrayFree)); }));
+      }
+      SECTION("ArrayDestroy") {
+        threads.push_back(std::thread(
+            [ptr]() { HIP_CHECK_THREAD(workIsDoneCheck<hipArray_t>(ptr, ArrayDestroy)); }));
+      }
+    }
+    for (auto& t : threads) {
+      t.join();
+    }
+    HIP_CHECK_THREAD_FINALIZE();
+  }
+}
+#else
+
+TEMPLATE_TEST_CASE("Unit_hipFreeMultiTArray", "", char, int, float2, float4) {
+  constexpr size_t numAllocs = 10;
+  std::vector<hipArray_t> ptrs(numAllocs);
+  hipExtent extent{};
+  extent.width = GENERATE(32, 128, 256, 512, 1024);
+  extent.height = GENERATE(0, 32, 128, 256, 512, 1024);
+  hipChannelFormatDesc desc = hipCreateChannelDesc<TestType>();
+
+  for (auto& ptr : ptrs) {
+    HIP_CHECK(hipMallocArray(&ptr, &desc, extent.width, extent.height, hipArrayDefault));
+  }
+
+  std::vector<std::thread> threads;
+
+  for (auto ptr : ptrs) {
+    SECTION("ArrayFree") {
+      threads.push_back(
+          std::thread([ptr]() { HIP_CHECK_THREAD(workIsDoneCheck<hipArray_t>(ptr, ArrayFree)); }));
+    }
+    SECTION("ArrayDestroy") {
```

Comment:
hipArrayDestroy should be used with hipArrayCreate

Reply:
Amended